src/metatensor/models/utils/data/dataset.py

```diff
@@ -109,30 +105,32 @@ def collate_fn(batch: List[NamedTuple]) -> Tuple[List, Dict[str, TensorMap]]:
     return systems, collated_targets
 
 
-def check_datasets(
-    train_datasets: List[Dataset],
-    validation_datasets: List[Dataset],
-    raise_incompatibility_error: bool = True,
-):
-    """
-    This is a helper function that checks that the training and validation sets
-    are compatible with one another.
+def check_datasets(train_datasets: List[Dataset], validation_datasets: List[Dataset]):
+    """Check that the training and validation sets are compatible with one another
 
     Although these checks will not fit all use cases, most models would be expected
-    to be able to use this function. If the validation set contains chemical species
-    or targets that are not present in the training set, this function will raise a
-    warning or an error, depending on the ``raise_incompatibility_error`` flag.
-
-    The option to warn is intended for model fine tuning, where a species or target
-    in the validation set might not be present in the current training set, but it
-    might have been present in the training of the base model.
-
-    :param train_datasets: A list of training datasets.
-    :param validation_datasets: A list of validation datasets.
-    :param raise_incompatibility_error: Whether to error (if ``true``) or warn
-        (if ``false``) upon detection of a chemical species or target in the
-        validation set that is not present in the training set.
+    to be able to use this function.
+
+    :param train_datasets: A list of training datasets to check.
+    :param validation_datasets: A list of validation datasets to check
+    :raises TypeError: If the ``dtype`` within the datasets are inconsistent.
+    :raises ValueError: If the `validation_datasets` has a target that is not present in
+        the ``train_datasets``.
+    :raises ValueError: If the training or validation set contains chemical species
+        or targets that are not present in the training set
     """
+    # Check that system `dtypes` are consistent within datasets
+    desired_dtype = train_datasets[0][0].system.positions.dtype
+    msg = f"`dtype` between datasets is inconsistent, found {desired_dtype} and "
+    for train_dataset in train_datasets:
+        actual_dtype = train_dataset[0].system.positions.dtype
+        if actual_dtype != desired_dtype:
+            raise TypeError(f"{msg}{actual_dtype} found in `train_datasets`")
+
+    for validation_dataset in validation_datasets:
+        actual_dtype = validation_dataset[0].system.positions.dtype
+        if actual_dtype != desired_dtype:
+            raise TypeError(f"{msg}{actual_dtype} found in `validation_datasets`")
```

Comment:
Should we do this? I think it's up to us to feed in the datasets with the correct dtype, so it wouldn't be a mistake on the side of the user or the architecture (note that this would show up as an architecture error by the way)

Reply:
We can leave this out, but usually I follow the convention that everything which is public should be able to be used independently from anyone. Following this I think it makes sens to test for the dtype even if I agree that the error should not happen of we do everything correct. Which in the past turned out not always to be true.

Reply:
You're right, we don't want to close the possibility for someone to use each model's trainer independently of metatensor-models.